service/raft/raft_group_registry.cc

```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
 
 raft_group_registry::~raft_group_registry() = default;
 
+namespace {
+    class composite_abort_source {
+        abort_source _as;
+        utils::small_vector<abort_source::subscription, 2> _subscriptions;
+    public:
+        void add(abort_source& as) {
+            as.check();
```

Comment:
Yes, my approach also wouldn't execute the code below (i.e. subscribe), but take alternative path: if `as` is aborted, then immediately `_as.request_abort()` (but don't throw from `add` itself)

Reply:
> This doesn't prevent the code after composite_abort_source::add from executing, throwing exception does.


Yeah, I guess the correct pattern would be then to do `check()` on the composite after `add()`. Throwing from `add()` indeed might be more convenient. Let's say you create a composite and then do a sequence of `add()`s, your way guarantees that the composite is not aborted.

On the other hand, let's say you create composite, `add()`, then yield, and `add()` again. You should still check the composite after yielding because the first `add()`ed abort_source might be aborted by this point...

Perhaps `add()` should also `check()` the composite at the end, if we take your approach?

Reply:
> > This doesn't prevent the code after composite_abort_source::add from executing, throwing exception does.
> 
> Yeah, I guess the correct pattern would be then to do `check()` on the composite after `add()`. Throwing from `add()` indeed might be more convenient. Let's say you create a composite and then do a sequence of `add()`s, your way guarantees that the composite is not aborted.
> 
> On the other hand, let's say you create composite, `add()`, then yield, and `add()` again. You should still check the composite after yielding because the first `add()`ed abort_source might be aborted by this point...
> 
> Perhaps `add()` should also `check()` the composite at the end, if we take your approach?

Checking on `_as` seems weird to me...  Let's leave it as is for now.